unconstrained_test.go

```diff
@@ -901,130 +1202,115 @@ func TestGradientDescentBacktracking(t *testing.T) {
+			t.Errorf("nil result without error for:\n%v", test)
 			continue
 		}
-		if result == nil {
-			t.Errorf("Case %v: nil result without error", i)
+
+		// Check that the optimum function value is as expected.
+		if math.Abs(result.F-test.optVal) > test.tol {
+			t.Errorf("Minimum not found, exited with status: %v. Want: %v, Got: %v for:\n%v",
+				result.Status, test.optVal, result.F, test)
 			continue
 		}
 
-		// rerun it again to ensure it gets the same answer with the same starting
-		// condition
-		result2, err2 := Local(test.F, test.X, test.Settings, method)
+		funcs, funcInfo := getFunctionInfo(test.f)
+		if funcInfo.IsGradient {
+			settings.InitialGradient = resize(settings.InitialGradient, len(test.x))
+			settings.InitialFunctionValue = funcs.function.F(test.x)
+			funcs.gradient.Df(test.x, settings.InitialGradient)
+			settings.UseInitialData = true
+		} else if funcInfo.IsFunctionGradient {
+			settings.InitialGradient = resize(settings.InitialGradient, len(test.x))
+			settings.InitialFunctionValue = funcs.gradFunc.FDf(test.x, settings.InitialGradient)
+			settings.UseInitialData = true
+		}
+
+		// Rerun the test again to make sure that it gets the same answer with
+		// the same starting condition.
+		result2, err2 := Local(test.f, test.x, settings, method)
 		if err2 != nil {
-			t.Errorf("error finding minimum second time: %v", err2.Error())
+			t.Errorf("error finding minimum second time %v for:\n%v", err2.Error(), test)
 			continue
 		}
+
 		if result2 == nil {
-			t.Errorf("Case %v: nil result without error", i)
+			t.Errorf("second time nil result without error for:\n%v", test)
 			continue
 		}
-		/*
-			// For debugging purposes, can't use DeepEqual naively becaus of NaNs
-			// kill the runtime before the check, because those don't need to be equal
-			result.Runtime = 0
-			result2.Runtime = 0
-			if !reflect.DeepEqual(result, result2) {
-				t.Error(eqString)
+
+		// At the moment all the optimizers are deterministic, so check that we
+		// get _exactly_ the same answer second time as well.
+		if result.F != result2.F {
+			t.Errorf("Different minimum second time. First: %v, Second: %v for:\n%v",
+				result.F, result2.F, test)
+		}
+
+		// Check that providing initial data reduces the number of function
+		// and/or gradient calls exactly by one.
+		if funcInfo.IsGradient {
+			if result.FunctionEvals != result2.FunctionEvals+1 {
+				t.Errorf("Providing initial data does not reduce the number of functions calls for:\n%v", test)
+				continue
+			}
+			if result.GradientEvals != result2.GradientEvals+1 {
+				t.Errorf("Providing initial data does not reduce the number of gradient calls for:\n%v", test)
 				continue
 			}
-		*/
+		} else if funcInfo.IsFunctionGradient {
+			if result.FunctionGradientEvals != result2.FunctionGradientEvals+1 {
+				t.Errorf("Providing initial data does not reduce the number of function/gradient calls for:\n%v", test)
+				continue
+			}
+		}
+
+		// if test.optLoc != nil {
```

Comment:
Optimizing from a (nearly) optimal location should successfully return either immediately or after a few iterations. However, due to the floating point effects the current implementation cannot handle many of the tests, either an error is returned or the call to Local does not even return. This is serious and must be addressed somehow, so this is commented out to serve as a kind of TODO.